File: frontend/src/lib/report-figure-catalog.ts
import sharedReportFigures from "../../../shared/report-figures.json" with { type: "json" };

export type ReportFigureScope = "control" | "all";

export interface ScopedReportFigure {
  scope: ReportFigureScope;
}

export interface BeamReportFigure extends ScopedReportFigure {
  imageKey: string;
  metric: "moment" | "shear" | "deflection";
  title: string;
  seriesLabel: string;
  unit: string;
}

export interface FrameMemberReportFigure extends ScopedReportFigure {
  overlayImageKey: string;
  traditionalImageKey: string;
  metric: "momentKnM" | "shearKn" | "deflectionMm" | "axialKn";
  label: string;
  title: string;
  unit: string;
}

export interface TrussReportFigure extends ScopedReportFigure {
  imageKey: string;
  metric: "axial" | "displacement";
  title: string;
  seriesLabel: string;
  unit: string;
}

interface SharedReportFigureCatalog {
  beam: {
    overlay: BeamReportFigure[];
    traditional: BeamReportFigure[];
  };
  frame: {
    member: FrameMemberReportFigure[];
  };
  truss: {
    overlay: TrussReportFigure[];
    traditional: TrussReportFigure[];
  };
}

const REPORT_FIGURES = sharedReportFigures as SharedReportFigureCatalog;

export const BEAM_REPORT_OVERLAY_FIGURES: readonly BeamReportFigure[] = REPORT_FIGURES.beam.overlay;
export const BEAM_REPORT_TRADITIONAL_FIGURES: readonly BeamReportFigure[] = REPORT_FIGURES.beam.traditional;
export const FRAME_REPORT_MEMBER_FIGURES: readonly FrameMemberReportFigure[] = REPORT_FIGURES.frame.member;
export const TRUSS_REPORT_OVERLAY_FIGURES: readonly TrussReportFigure[] = REPORT_FIGURES.truss.overlay;
export const TRUSS_REPORT_TRADITIONAL_FIGURES: readonly TrussReportFigure[] = REPORT_FIGURES.truss.traditional;

export function reportFiguresForScope<T extends ScopedReportFigure>(figures: readonly T[], includeAll: boolean): T[] {
  return figures.filter((figure) => includeAll || figure.scope === "control");
}

File: backend/exporters/common/report_figure_catalog.py
from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Literal, TypeVar

ReportFigureScope = Literal["control", "all"]
TReportFigure = TypeVar("TReportFigure")
_REPORT_FIGURES_PATH = Path(__file__).resolve().parents[3] / "shared" / "report-figures.json"


@dataclass(frozen=True)
class BeamReportFigure:
    image_key: str
    metric: Literal["moment", "shear", "deflection"]
    title: str
    series_label: str
    unit: str
    scope: ReportFigureScope


@dataclass(frozen=True)
class FrameMemberReportFigure:
    overlay_image_key: str
    metric_key: Literal["momentKnM", "shearKn", "deflectionMm", "axialKn"]
    label: str
    title: str
    unit: str
    scope: ReportFigureScope


@dataclass(frozen=True)
class TrussReportFigure:
    image_key: str
    metric: Literal["axial", "displacement"]
    title: str
    series_label: str
    unit: str
    scope: ReportFigureScope


@lru_cache(maxsize=1)
def _report_figure_catalog() -> dict:
    return json.loads(_REPORT_FIGURES_PATH.read_text(encoding="utf-8"))


def _beam_figure(row: dict) -> BeamReportFigure:
    return BeamReportFigure(
        image_key=str(row["imageKey"]),
        metric=str(row["metric"]),  # type: ignore[arg-type]
        title=str(row["title"]),
        series_label=str(row["seriesLabel"]),
        unit=str(row["unit"]),
        scope=str(row["scope"]),  # type: ignore[arg-type]
    )


def _frame_member_figure(row: dict) -> FrameMemberReportFigure:
    return FrameMemberReportFigure(
        overlay_image_key=str(row["overlayImageKey"]),
        metric_key=str(row["metric"]),  # type: ignore[arg-type]
        label=str(row["label"]),
        title=str(row["title"]),
        unit=str(row["unit"]),
        scope=str(row["scope"]),  # type: ignore[arg-type]
    )


def _truss_figure(row: dict) -> TrussReportFigure:
    return TrussReportFigure(
        image_key=str(row["imageKey"]),
        metric=str(row["metric"]),  # type: ignore[arg-type]
        title=str(row["title"]),
        series_label=str(row["seriesLabel"]),
        unit=str(row["unit"]),
        scope=str(row["scope"]),  # type: ignore[arg-type]
    )


BEAM_REPORT_OVERLAY_FIGURES: tuple[BeamReportFigure, ...] = tuple(_beam_figure(row) for row in _report_figure_catalog()["beam"]["overlay"])
BEAM_REPORT_TRADITIONAL_FIGURES: tuple[BeamReportFigure, ...] = tuple(_beam_figure(row) for row in _report_figure_catalog()["beam"]["traditional"])
FRAME_REPORT_MEMBER_FIGURES: tuple[FrameMemberReportFigure, ...] = tuple(_frame_member_figure(row) for row in _report_figure_catalog()["frame"]["member"])
TRUSS_REPORT_OVERLAY_FIGURES: tuple[TrussReportFigure, ...] = tuple(_truss_figure(row) for row in _report_figure_catalog()["truss"]["overlay"])


def report_figures_for_scope(figures: tuple[TReportFigure, ...], include_all: bool) -> tuple[TReportFigure, ...]:
    if include_all:
        return figures
    return tuple(figure for figure in figures if getattr(figure, "scope", "all") == "control")
